Support generic row (variant) keys in VSM

cseed opened the pull request.

tpoterba commented on src/main/scala/is/hail/methods/DuplicateAccumulator.scala:
```diff
@@ -1,56 +0,0 @@
-
-import scala.collection.mutable
-
-object DuplicateAccumulator extends AccumulableParam[(Long, mutable.Set[Variant]), Variant] {
```
discussed this. OK

tpoterba commented on src/main/scala/is/hail/variant/GenericDatasetFunctions.scala:
```diff
@@ -0,0 +1,106 @@
+package is.hail.variant
```
I'm not looking at this because it's just a direct copy, right?

cseed replied:
I don't understand.

tpoterba commented on src/main/scala/is/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -40,81 +42,95 @@ object VariantSampleMatrix {
+          val v = r.getVariant(0)
+          (v, (vaImporter(r.get(1)),
+            if (dropSamples)
+              Iterable.empty[Genotype]
```
This should use `GenotypeStream.empty()` given my discovery about the cost of not doing anything :man_shrugging:

tpoterba commented on src/main/scala/is/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -248,8 +264,9 @@ object VariantSampleMatrix {
-        case JObject(List(("id", JString(id)), ("annotation", jv: JValue))) =>
-          (id, JSONAnnotationImpex.importAnnotation(jv, saSignature, "sample_annotations"))
+        case JObject(List(("id", id), ("annotation", jv))) =>
+          (JSONAnnotationImpex.importAnnotation(id, sSignature, "sample_annotations.id"),
```
should these be `column_keys` and `column_annotations`?

cseed replied:
No, they come from the "sample_annotations" field of the metadata JSON.  With this change, I tried not to rename anything related to sample or variant to avoid too much chaos.  The only exception is rowkey_schem and colkey_schema on python VariantDataset since variant_schema and sample_schema were already taken.

tpoterba commented on src/main/scala/is/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -2106,7 +2118,128 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VSMMetadata,
+        fatal(s"unknown StorageLevel `$storageLevel'")
+    }
+
+    copy(rdd = rdd.persist(level))
```
We've lost the genotypeStream optimization here. That was pretty important - if you do a `filter_samples` to 1% of your samples and then persist with this method, you end up storing everything on disk given the lazy filter implementation

cseed replied:
Oops, good catch.

cseed commented on src/main/scala/is/hail/expr/AnnotationImpex.scala:
```diff
@@ -53,6 +53,13 @@ object SparkAnnotationImpex extends AnnotationImpex[DataType, Any] {
         }: _*)
   }
 
+  def annotationImporter(t: Type): (Any) => Annotation = {
```
Yes, I did a version of this, but led to a high number of composed, higher-ordered functions (like we had with the AST eval method) and it didn't seem better.  But we could use asm to generic a Type-specific importer, that would be nice.

cseed commented:
Addressed comments and rebased.  VSM still has RPK, I'm still studying if removing it is feasible.

cseed commented:
@tpoterba ready for another look.

I killed isGenericGenotype.  filterGenotypes just makes filtered cells null.  GenotypeStream now supports null genotypes, this is a breaking change.  I bumped the file version (to 0x101) and nuked the 0.1 and other legacy tests.

The merge target is now the 0.2 branch.  Once the CI is correctly handling 0.1 and devel/0.2, we can merge 0.2 into the master and just work out of master.

cseed commented:
Allow the Genotype in VSM to be null.  This means you can't call `g.gt` anymore, since `g` might be null.  Moved the user-visible Genotype functions to the Genotype companion object and made them null-safe.  This makes the behavior of filterGenotypes between VariantDataset and GenericDataset consistent.  Fixed the tests.  @tpoterba ready for another look.

cseed commented:
Fixed all the tests.

tpoterba commented on python/hail/docs/conf.py:
```diff
@@ -72,7 +72,43 @@
 
 hc = HailContext(log="output/hail.log", quiet=True)
-vds = hc.read('data/example.vds')"""
+
```
why is this stuff in conf and not a test?

tpoterba replied:
ok, sounds fine

tpoterba commented on src/main/scala/is/hail/io/vcf/ExportVCF.scala:
```diff
@@ -150,27 +150,33 @@ object ExportVCF {
   def writeGenotype(sb: StringBuilder, g: Genotype) {
-    sb.append(g.gt.map { gt =>
+    if (g == null) {
+      sb.append("./.")
```
does this compile to the same thing as if we saved this "./." string as a val somewhere on a static object?

cseed replied:
Yes.  String constants are stored in the Java Constant Pool.

tpoterba replied:
great, thanks!

tpoterba commented on src/main/scala/is/hail/methods/FilterAlleles.scala:
```diff
@@ -120,19 +120,22 @@ object FilterAlleles {
+          null
+        else {
+          val px = Genotype.px(g).map(subsetPx)
+          g.copy(
```
this code is somewhat horrifying but we'll remove it soon enough

tpoterba commented on src/main/scala/is/hail/methods/SampleQC.scala:
```diff
@@ -71,53 +71,57 @@ class SampleQCCombiner(val keepStar: Boolean) extends Serializable {
+    if (g == null)
+      nNotCalled += 1
+    else
+      Genotype.gt(g) match {
```
can we make this unsafeGT?

tpoterba commented on src/main/scala/is/hail/methods/SplitMulti.scala:
```diff
@@ -44,16 +44,17 @@ object SplitMulti {
     for (g <- it) {
-
-      val gadsum = g.ad.map(gadx => (gadx, gadx.sum))
+      val gadsum = Genotype.ad(g).map(gadx => (gadx, gadx.sum))
```
I am very triggered.

cseed replied:
Me too, me too.  I basically think we should rip out Genotype and then worry about how to speed up generic genotypes.

tpoterba commented on src/main/scala/is/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -40,81 +42,95 @@ object VariantSampleMatrix {
+              implicit val pkjr = vSignature.partitionKey.jsonReader
+              jv.fromJSON[OrderedPartitioner[Annotation, Annotation]]
+            } catch {
+              case _: FileNotFoundException =>
```
we've bumped the file format. remove this

tpoterba commented on src/main/scala/is/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -40,81 +42,95 @@ object VariantSampleMatrix {
+            jv.fromJSON[OrderedPartitioner[Locus, Variant]]
+          } catch {
+            case _: FileNotFoundException =>
+              fatal("missing partitioner.json.gz when loading VDS, create with HailContext.write_partitioning.")
```
remove

cseed commented:
@tpoterba Back to you.  It's a brave new world.
